src/handler/sign_tx.c

```diff
@@ -1,103 +1,423 @@
+
+
+bool sign_tx_with_key() {
+    // the bip32 path and bip32 path length should be on the global context when calling this method.
+
+
+    cx_ecfp_private_key_t private_key = {0};
+    cx_ecfp_public_key_t public_key = {0};
+
+    uint8_t chain_code[32];
+    uint8_t hash_160[20];
+
+    derive_private_key(&private_key, chain_code, G_context.bip32_path.path, G_context.bip32_path.length);
+    init_public_key(&private_key, &public_key);
+
+    if (G_context.tx_info.sighash_all[0] == '\0') {
+        // finish sha256 from data
+        cx_hash(&G_context.tx_info.sha256.header, CX_LAST, G_context.tx_info.sighash_all, 0, G_context.tx_info.sighash_all, 32);
+        cx_sha256_init(&G_context.tx_info.sha256);
+        cx_hash(&G_context.tx_info.sha256.header, CX_LAST, G_context.tx_info.sighash_all, 32, G_context.tx_info.sighash_all, 32);
+        // now get second sha256
+    }
+
+    uint8_t out[256] = {0};
+    size_t sig_size = cx_ecdsa_sign(&private_key, CX_LAST | CX_RND_RFC6979, CX_SHA256, G_context.tx_info.sighash_all, 32, out, 256, NULL);
+
+    explicit_bzero(&private_key, sizeof(private_key));
+    explicit_bzero(&public_key, sizeof(public_key));
+
+    // exchange signature
+    // io_send_response < 0 means faillure
+    return io_send_response(&(const buffer_t){.ptr = out, .size = sig_size, .offset = 0}, SW_OK) >= 0;
+}
+
+void _decode_next_element() {
+    // test if there are more tokens, inputs and outputs in this order
+    // if there are, try to read from buffer
+    // if any are incomplete, inform caller to send more data 
+    if (G_context.tx_info.remaining_tokens > 0) {
+        // can read token?
+        if (G_context.tx_info.buffer_len < TOKEN_UID_LEN) {
+            // still decoding tokens, but this one was divided between calls 
+            THROW(TX_STATE_PARTIAL);
+        }
+        // for now we ignore it
+        G_context.tx_info.remaining_tokens--;
+        G_context.tx_info.buffer_len -= TOKEN_UID_LEN;
+        G_context.tx_info.elem_type = ELEM_TOKEN_UID;
+        memmove(G_context.tx_info.buffer, G_context.tx_info.buffer + TOKEN_UID_LEN, G_context.tx_info.buffer_len);
+    } else if(G_context.tx_info.remaining_inputs > 0) {
+        // read input
+        if (G_context.tx_info.buffer_len < TX_INPUT_LEN) {
+            THROW(TX_STATE_PARTIAL);
+        }
+        uint16_t input_val = read_u16_be(G_context.tx_info.buffer, 33);
+        if (input_val>0) {
+            // we require input to have no data, since we are signing all data received (sighash_all must have no data)
+            THROW(TX_STATE_ERR);
         }
+        // reading input
+        G_context.tx_info.remaining_inputs--;
+        G_context.tx_info.buffer_len -= TX_INPUT_LEN;
+        G_context.tx_info.elem_type = ELEM_INPUT;
+        memmove(G_context.tx_info.buffer, G_context.tx_info.buffer + TX_INPUT_LEN, G_context.tx_info.buffer_len);
+    } else if (G_context.tx_info.current_output < G_context.tx_info.outputs_len) {
+        // read output
```

Comment:
Can you add a brief explanation of this case?

Reply:
My point here is: What does it mean to enter in this `else if` block?

Reply:
Commented this section explaining what each step does, but jus entering this `else if` means we have at least 1 more output to read.

Reply:
Thanks!